src/symmetric_crypto.cpp

```diff
@@ -0,0 +1,382 @@
+                }
+            } break;
+        }
+
+        openssl::_EVP_CipherInit_ex(_ctx.get(),
+                                    EVPCipherConstructor(),
+                                    nullptr,
+                                    secretKey.data(),
+                                    iv.data(),
+                                    this->_operation == Operation::Encryption);
+
+        switch (padding) {
+            case SymmetricCipherPadding::PKCS:
+                // OpenSSL uses PKCS padding by default.
+                break;
+        }
+    };
+
+    void update(const std::vector<uint8_t> &plaintext)
+    {
+        if (_isFinished) {
+            throw MoCOCrWException(
+                    "Further calls to update() are not allowed once finish() was called.");
+        }
+
+        int processingChunkSize = plaintext.size();
+        std::vector<uint8_t> processedChunk(processingChunkSize + EVP_MAX_BLOCK_LENGTH);
+
+        openssl::_EVP_CipherUpdate(_ctx.get(),
+                                   processedChunk.data(),
+                                   &processingChunkSize,
+                                   plaintext.data(),
+                                   plaintext.size());
+
+        processedChunk.resize(processingChunkSize);
+        _bufferStrategy->write(processedChunk);
+    }
+
+    std::vector<uint8_t> read(size_t length)
+    {
+        return _bufferStrategy->read(length);
+    }
+
+    std::vector<uint8_t> readAll()
+    {
+        return _bufferStrategy->readAll();
+    }
+
+    std::vector<uint8_t> finish()
+    {
+        if (_isFinished) {
+            throw MoCOCrWException("finish() can't be called twice.");
+        }
+
+        int processingChunkSize = 0;
+        std::vector<uint8_t> processedChunk(EVP_MAX_BLOCK_LENGTH);
+
+        if (isAuthenticatedCipherMode(_mode) && _operation == Operation::Decryption) {
+            if (_authTag.size() != AuthenticatedAESCipher::AuthTagLength) {
+                throw MoCOCrWException("Authentication Tag must be set before calling finish() "
+                                       "for cipher which implements authenticated encryption.");
+            }
+            openssl::_EVP_CIPHER_CTX_ctrl(_ctx.get(), EVP_CTRL_GCM_SET_TAG,
+                                          AuthenticatedAESCipher::AuthTagLength, _authTag.data());
+        }
+
+        openssl::_EVP_CipherFinal_ex(_ctx.get(), processedChunk.data(), &processingChunkSize);
+
+        if (_mode == SymmetricCipherMode::GCM && _operation == Operation::Encryption) {
+            _authTag.resize(AuthenticatedAESCipher::AuthTagLength);
+            openssl::_EVP_CIPHER_CTX_ctrl(_ctx.get(), EVP_CTRL_GCM_GET_TAG,
+                                          AuthenticatedAESCipher::AuthTagLength,
+                                          _authTag.data());
+        }
+
+        processedChunk.resize(processingChunkSize);
+        _bufferStrategy->write(processedChunk);
```

Comment:
Same here, this should probably be `std::move(processedChunk)`

Reply:
fixed